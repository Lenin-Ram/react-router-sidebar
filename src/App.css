.routeDemo {
  color: white;
  font-weight: bold;
  background-color: #4b4ba3e3;
  text-shadow: 10px 10px #0000006b;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 5em;
}
