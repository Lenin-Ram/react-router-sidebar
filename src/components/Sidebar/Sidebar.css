.sidebar {
  background-color: #fafafa;
  height: 100vh;
  border-right: 1px solid #dbdbdb;
  position: fixed;
  transition: width 300ms ease-in-out 0s;
}

.collapse {
  display: flex;
  position: fixed;
  bottom: 0;
  border-top: 1px solid #dbdbdb;
  cursor: pointer;
  background-color: #fafafa;
  font-size: 0.875rem;
  color: #666666;
  text-transform: uppercase;
  height: 48px;
  justify-content: space-evenly;
  align-items: center;
  transition: width 300ms ease-in-out 0s;
}

.collapse:hover {
  background-color: #dbdbdb;
}

.collapse > i {
  flex: 1;
  text-align: center;
  animation: fadeIn 800ms ease 100ms 1 normal both;
  font-size: 20px;
}
.collapse > p {
  animation: fadeIn 800ms ease 100ms 1 normal both;
  flex: 3;
}

@keyframes shrinkAvatar {
  0% {
    width: 10rem;
    height: 10rem;
  }
  100% {
    width: 3rem;
    height: 3rem;
  }
}

@keyframes expandAvatar {
  0% {
    width: 3rem;
    height: 3rem;
  }
  100% {
    width: 10rem;
    height: 10rem;
  }
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
