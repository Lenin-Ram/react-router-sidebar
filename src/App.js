import React from 'react'
import { Link, Route, Switch } from 'react-router-dom'
import './App.css'
import Estadistica from './components/estadistica/estadistica';
import Formulario from "./components/formulario/formulario";
import Sidebar from './components/Sidebar/Sidebar'

export const sideBarObject = {
  header: 'URL Shortener',
  avatar: 'http://link-butcher-2.herokuapp.com/MDE2YzE2',
  routeTree: [
    {
      children: [
        { name: 'Inicio', route: '/dashboard/r' },
        { name: 'Estadistica', route: '/dashboard/o' },
        { name: 'Formulario', route: '/dashboard/f' },
      ],
      icon: 'home',
      name: 'Menu',
    },
   
    
  ],
}

function RouteDemo({ children }) {
  return <div className="routeDemo">{`You are here > ${children}`}</div>
}

function App() {
  return (
    <div className="App">
      <Switch>
        <Route path="/dashboard" component={Dashboard} />
        <Route path="/" component={Login}></Route>
      </Switch>
    </div>
  )
}

function Login() {
  return <Link to="/dashboard">Click here to Login</Link>
}

function Dashboard({ match }) {
  return (
    <div>
      <div>
        <Sidebar
          sideBarObject={sideBarObject}
          initialRoute={window.location.pathname}
        ></Sidebar>
      </div>
      <Switch>
        <Route path={match.url + '/o'}
        // component={Estadistica}
        >
          
        </Route>

        <Route path={match.url + '/f'}
        component={Formulario}>
 
        </Route>
        <Route path={match.url + '/'}>
          <RouteDemo>Home</RouteDemo>
        </Route>
      </Switch>
    </div>
  )
}
export default App
