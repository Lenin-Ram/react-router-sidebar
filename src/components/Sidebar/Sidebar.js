import React, { useEffect, useState } from 'react'
import { Link, NavLink } from 'react-router-dom'
import './Sidebar.css'

function SidebarRoot({ routeTuple, headerOnCLick, collapsed, avatar }) {
  return (
    <NavLink
      className="sidebar-root"
      activeClassName="active"
      to={routeTuple.route}
      onClick={() => {
        headerOnCLick(routeTuple.route)
      }}
      style={collapsed ? { alignItems: 'center' } : {}}
    >
      <div
        style={
          collapsed
            ? {
                animationName: 'shrinkAvatar',
                backgroundImage: `url(${avatar})`,
              }
            : {
                animationName: 'expandAvatar',
                backgroundImage: `url(${avatar})`,
              }
        }
      ></div>
      <p
        style={
          collapsed
            ? { display: 'none' }
            : {
                padding: '10px 16px',
              }
        }
      >
        {routeTuple.name}
      </p>
    </NavLink>
  )
}

function Tree({ routeTree, showTree, headerOnCLick, collapsed }) {
  const [showFlyout, setShowFlyout] = useState(false)
  return (
    <div className={`sidebar-tree ${showTree ? 'active' : ''}`}>
      <Link
        to={routeTree.children[0].route}
        onClick={() => {
          headerOnCLick()
        }}
        className="header-link"
        onMouseEnter={() => {
          setShowFlyout(true)
        }}
        onMouseLeave={() => {
          setTimeout(() => {
            setShowFlyout(false)
          }, 50)
        }}
        style={{
          color: showTree ? '#4b4ba3' : '#666',
          fontWeight: showTree ? 'bold' : 'normal',
          flex: 1,
        }}
      >
        {routeTree.icon && (
          <i
            className={`fas fa-${routeTree.icon}`}
            style={
              collapsed
                ? {
                    flex: 1,
                    textAlign: 'center',
                  }
                : {}
            }
          ></i>
        )}
        {!collapsed && <p>{routeTree.name}</p>}
      </Link>

      <div
        className={
          showTree && !collapsed
            ? 'sub'
            : !collapsed
            ? 'flyout'
            : 'flyout collapsed'
        }
        style={
          showFlyout && (!showTree || collapsed) ? { display: 'flex' } : {}
        }
      >
        {routeTree.children.map((routeTuple, index) => (
          <NavLink
            key={index}
            exact
            to={routeTuple.route}
            activeClassName="active"
            onClick={() => {
              headerOnCLick()
            }}
          >
            {routeTuple.name}
          </NavLink>
        ))}
      </div>
    </div>
  )
}

export default function ({ sideBarObject, initialRoute }) {
  const [indexOfActiveTreeHeader, setIndexOfActiveTreeHeader] = useState(null)
  const [isCollapsed, setIsCollapsed] = useState(false)
  function rebuildTree(auxRoute) {
    sideBarObject.routeTree.forEach((routeTreeObject, index) => {
      routeTreeObject.children.forEach((routeTuple) => {
        if (
          (auxRoute && routeTuple.route === auxRoute) ||
          routeTuple.route === initialRoute
        ) {
          setIndexOfActiveTreeHeader(index)
        }
      })
    })
  }
  useEffect(rebuildTree, [])

  return (
    <div
      className="sidebar"
      style={{
        width: isCollapsed ? '4rem' : '15rem',
      }}
    >
      <SidebarRoot
        avatar={sideBarObject.avatar}
        routeTuple={{
          name: sideBarObject.header,
          route: sideBarObject.routeTree[0].children[0].route,
        }}
        headerOnCLick={rebuildTree}
        collapsed={isCollapsed}
      ></SidebarRoot>
      {sideBarObject.routeTree.map((routeTreeObject, index) => (
        <Tree
          collapsed={isCollapsed}
          key={index}
          routeTree={routeTreeObject}
          showTree={index === indexOfActiveTreeHeader}
          headerOnCLick={() => {
            setIndexOfActiveTreeHeader(index)
          }}
        ></Tree>
      ))}
      <div
        className="collapse"
        onClick={() => {
          setIsCollapsed(!isCollapsed)
        }}
        style={
          isCollapsed
            ? {
                width: '4rem',
              }
            : {
                width: '15rem',
              }
        }
      >
        <i
          className={
            isCollapsed
              ? 'fas fa-angle-double-right'
              : 'fas fa-angle-double-left'
          }
        ></i>
        <p style={isCollapsed ? { display: 'none' } : { display: 'flex' }}>
          Collapse sidebar
        </p>
      </div>
    </div>
  )
}
