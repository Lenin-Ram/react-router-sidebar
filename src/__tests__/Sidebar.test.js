import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { sideBarObject } from '../App'
import Sidebar from '../components/Sidebar/Sidebar'

it('renders the sample route tree correctly', () => {
  const DOMTree = renderer
    .create(
      <Router>
        <Sidebar sideBarObject={sideBarObject} initialRoute="/" />
      </Router>
    )
    .toJSON()
  expect(DOMTree).toMatchSnapshot()
})
