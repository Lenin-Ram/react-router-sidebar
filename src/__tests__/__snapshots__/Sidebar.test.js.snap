// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the sample route tree correctly 1`] = `
<div
  className="sidebar"
  style={
    Object {
      "width": "15rem",
    }
  }
>
  <a
    aria-current="page"
    className="sidebar-root active"
    href="/"
    onClick={[Function]}
    style={Object {}}
  >
    <div
      style={
        Object {
          "animationName": "expandAvatar",
          "backgroundImage": "url(http://link-butcher-2.herokuapp.com/MDE2YzE2)",
        }
      }
    />
    <p
      style={
        Object {
          "padding": "10px 16px",
        }
      }
    >
      URL Shortener
    </p>
  </a>
  <div
    className="sidebar-tree "
  >
    <a
      className="header-link"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "color": "#666",
          "flex": 1,
          "fontWeight": "normal",
        }
      }
    >
      <i
        className="fas fa-home"
        style={Object {}}
      />
      <p>
        Project Overview
      </p>
    </a>
    <div
      className="flyout"
      style={Object {}}
    >
      <a
        aria-current="page"
        className="active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        Progress
      </a>
      <a
        aria-current={null}
        href="/r"
        onClick={[Function]}
      >
        Contributions
      </a>
      <a
        aria-current={null}
        href="/o"
        onClick={[Function]}
      >
        Activity
      </a>
    </div>
  </div>
  <div
    className="sidebar-tree "
  >
    <a
      className="header-link"
      href="/lo"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "color": "#666",
          "flex": 1,
          "fontWeight": "normal",
        }
      }
    >
      <i
        className="fas fa-burn"
        style={Object {}}
      />
      <p>
        Oil
      </p>
    </a>
    <div
      className="flyout"
      style={Object {}}
    >
      <a
        aria-current={null}
        href="/lo"
        onClick={[Function]}
      >
        Frytol
      </a>
      <a
        aria-current={null}
        href="/t"
        onClick={[Function]}
      >
        Eno Serwaa
      </a>
      <a
        aria-current={null}
        href="/v"
        onClick={[Function]}
      >
        Pepsi
      </a>
      <a
        aria-current={null}
        href="/u"
        onClick={[Function]}
      >
        Soyabean
      </a>
    </div>
  </div>
  <div
    className="sidebar-tree "
  >
    <a
      className="header-link"
      href="/g"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "color": "#666",
          "flex": 1,
          "fontWeight": "normal",
        }
      }
    >
      <i
        className="fas fa-brain"
        style={Object {}}
      />
      <p>
        Minds
      </p>
    </a>
    <div
      className="flyout"
      style={Object {}}
    >
      <a
        aria-current={null}
        href="/g"
        onClick={[Function]}
      >
        Sardine
      </a>
      <a
        aria-current={null}
        href="/j"
        onClick={[Function]}
      >
        DHall Beast
      </a>
      <a
        aria-current={null}
        href="/d"
        onClick={[Function]}
      >
        Last Killer
      </a>
      <a
        aria-current={null}
        href="/l"
        onClick={[Function]}
      >
        Brilla
      </a>
      <a
        aria-current={null}
        href="/e"
        onClick={[Function]}
      >
        Ant Brain
      </a>
    </div>
  </div>
  <div
    className="collapse"
    onClick={[Function]}
    style={
      Object {
        "width": "15rem",
      }
    }
  >
    <i
      className="fas fa-angle-double-left"
    />
    <p
      style={
        Object {
          "display": "flex",
        }
      }
    >
      Collapse sidebar
    </p>
  </div>
</div>
`;
